File: src/languages/finnish/finnish.ts
/**
 * Finnish (fi) phonemizer — Standard Finnish (yleiskieli), the Latin orthography, canonical IPA.
 * The national language of Finland (~5.4M). Finnish is one of the most PHONEMICALLY TRANSPARENT orthographies in the
 * world (very nearly one grapheme ↔ one phoneme), so this is a greedy longest-match scan over the grapheme table
 * (manifest.ts) with three code rules: CONSONANT GEMINATION — a doubled consonant → geminate [Cː] (kukka→kukːɑ,
 * tullut→tulːut) — and the velar-nasal rules ⟨ng⟩→ŋː (a LONG velar nasal: kengät→keŋːæt, rengas→reŋːɑs) and ⟨nk⟩→ŋk
 * (⟨n⟩→ŋ before k: sänky→sæŋky). Signatures: 8 vowels with ⟨a⟩=ɑ (BACK); DOUBLING = LENGTH (aa→ɑː …); the 18
 * diphthongs mark the 2nd vowel as the non-syllabic offglide (au→ɑu̯, uo→uo̯) in the table; ⟨v⟩=ʋ (approximant),
 * ⟨r⟩=r (trill), ⟨j⟩=j. Consonant GRADATION is already spelled out in the orthography → no gradation logic needed.
 * Fixed word-initial primary stress is predictable + unwritten → not emitted (folded in the referee eval).
 */
import type { Phonemizer } from "../../registry.ts";
import { assembleClauses } from "../../core/clauses.ts";
import { latinPhone } from "../../core/latinPhones.ts";
import { LATIN_RUN, makeNativiser } from "../../core/hostWord.ts";
import { makeSymbolNormalizer } from "../../core/normalizeSymbols.ts";
import { MANIFEST, GRAPHEME_KEYS } from "./manifest.ts";
import { numberToWords, readDigits } from "./numbers.ts";
import { normalizeFinnish, normalizeFinnishInitialisms } from "./normalize.ts";

const G = MANIFEST.graphemes;
const CLAUSE_MARK = MANIFEST.clausePunctuation;
const VOWEL_LETTERS = new Set(MANIFEST.vowelLetters);

/** Phonemize a single Finnish word to canonical IPA (segmental; gemination + velar-nasal rules; length + diphthong
 *  offglides emitted). */
export function phonemizeWord(word: string): string {
    const w = word.toLowerCase();
    let out = "";
    let i = 0;
    while (i < w.length) {
        const c = w[i]!;
        const next = w[i + 1] ?? "";
        // ⟨ng⟩ → ŋː (a long velar nasal), consuming both letters — before gemination so the ⟨n⟩ isn't mishandled.
        if (c === "n" && next === "g") { out += "ŋː"; i += 2; continue; }
        // ⟨nk⟩ → ŋk (⟨n⟩ → ŋ before a velar k); emit ŋ and let the k be scanned next (so kk-gemination still holds).
        if (c === "n" && next === "k") { out += "ŋ"; i += 1; continue; }
        // consonant gemination: a doubled consonant letter → geminate [Cː].
        if (!VOWEL_LETTERS.has(c) && next === c && G[c]) { out += G[c]! + "ː"; i += 2; continue; }
        let matched = false;
        for (const key of GRAPHEME_KEYS) {
            if (w.startsWith(key, i)) { out += G[key]!; i += key.length; matched = true; break; }
        }
        // ⚠ NOT SILENTLY: a letter with no grapheme rule here still denotes a sound, and dropping it deletes
        // what the writer typed. Consulted only on the MISS branch, after every grapheme (including every
        // digraph) has been tried, so it can never override a reading this language has an opinion about.
        if (!matched) {
            out += latinPhone(w[i]!, { initial: i === 0, includeH: true }) ?? "";
            i += 1;
        }
    }
    return out;
}

// A word (Finnish Latin letters incl. ä ö å + loan š ž) / number / punctuation token.
const TOKEN = new RegExp(`(${LATIN_RUN})|(\\d+)|([.!?…,;:])`, "giu");

/**
 * This language's OWN inventory. ⚠ TWO DIFFERENT QUESTIONS, KEPT APART: the TOKEN class above decides where
 * the SCRIPT boundary falls (routing), while this one decides whether the g2p has rules for these letters. A
 * token this class REJECTS carries a letter the language does not use — i.e. a foreign name. See
 * core/hostWord.ts.
 */
const NATIVE_CLASS = "[a-zäöåšž]";
const nat = makeNativiser(NATIVE_CLASS, "iu");

/**
 * The shared symbol tier (core/normalizeSymbols.ts), which composes number + sign / unit / rate /
 * exponent in one pass. Runs AFTER normalize.ts, which is what lets that file keep its rewrites in
 * DIGITS: `13,6 cm` leaves normalize.ts as `13 pilkku 6 cm`, so the number–unit adjacency this tier
 * matches on is still intact.
 *
 * COUNT FORMS ARE `[singular, partitive]` AND THE DEFAULT SELECTOR IS EXACTLY RIGHT FOR FINNISH — a
 * counted noun is nominative after 1 and PARTITIVE after everything else (*1 kilometri*, *10
 * kilometriä*), which is the tier's `n === 1 ? 0 : 1` without an override.
 *
 * ⚠ `km/h` AND `m/s` ARE COMPOUND KEYS, NOT `unitPer` (trap 44 / trap 47 reason 1). The Finnish rate
 * idiom is not "A per B": the denominator takes the INESSIVE and there is no joining word at all —
 * *kilometriä tunnissa*, *metriä sekunnissa*. `unitPer` is one invariant string and none of these is one.
 * `unitAlt` is sorted longest-first, so the 4-character key is tried before the bare `km`.
 *
 * ⚠ BARE `m` IS DECLARED, AND THE ONE-LETTER-KEY HAZARD (traps 28/46) WAS MEASURED FIRST: digit-adjacent
 * `m` is ×1 in the retained corpus (`jopa 10 m korkeita`, a genuine metre) and the `version-dot` cell is
 * ×0. The residual exposure is a dotted designation ending in ⟨m⟩ (`802.11m`), which the tier's
 * `NOT_VERSION` guard rejects by SEEING THE DOT — and unlike fa/ckb/af/ca, this language's normalize.ts
 * never spends a decimal POINT (Finnish decimals use a comma), so the dot is still there when the tier
 * runs. The guard's evidence outlives the pass, which is what makes the key safe here (trap 39).
 *
 * ⚠ `°C` IS NOT ON THE TIER — normalize.ts step 10 owns it, because it also has to read the BARE `°` of
 * a latitude and an angle, which is the same word (*astetta*) and no scale name.
 */
const SYMBOLS = makeSymbolNormalizer({
    percent: ["prosentti", "prosenttia"],
    currency: { $: ["dollari", "dollaria"], "€": ["euro", "euroa"] },
    units: {
        // Longest-first inside the tier; both rate keys compose the inessive denominator themselves.
        "km/h": ["kilometri tunnissa", "kilometriä tunnissa"],
        "m/s": ["metri sekunnissa", "metriä sekunnissa"],
        km: ["kilometri", "kilometriä"],
        cm: ["senttimetri", "senttimetriä"],
        mm: ["millimetri", "millimetriä"],
        kg: ["kilogramma", "kilogrammaa"],
        m: ["metri", "metriä"],
    },
    // Finnish welds the measure word onto the FRONT as one compound — *neliökilometriä*, *kuutiometriä* —
    // which is `compound`, never `before` (that would give *neliö kilometriä*, two tokens).
    // `neliökilometriä` is corpus-attested ×8 and `neliökilometrin` ×20/12 on the wiki; `kuutio` is a
    // referee lemma and the cube has ×0 corpus instances, so it is declared for the shape rather than for
    // a measured defect and is labelled as such.
    exponentWords: { squared: ["neliö"], cubed: ["kuutio"], position: "compound" },
    // The magnitude hop, for `1 850 miljardia dollaria` — and the field gates the UNIT path's connective
    // hop too, which is the "one declaration, two consumers" note at the end of the playbook. Finnish
    // takes NO connective (*viisi miljoonaa dollaria*), so `magnitudeConnective` stays undefined.
    magnitudes: ["miljoona", "miljoonaa", "miljardi", "miljardia", "biljoona", "biljoonaa"],
    ampersand: "ja",
    // The corpus's `×` is a dimension cross (`4 096 × 2 304`, `7680×4320`) and a multiplier
    // (`2×15 minuuttia`); Finnish says *kertaa* for both, so `by` defaults to `times`.
    multiply: { times: "kertaa" },
});

class FinnishPhonemizer implements Phonemizer {
    text(rawInput: string): string {
        const input = SYMBOLS(normalizeFinnishInitialisms(normalizeFinnish(rawInput)));
        return assembleClauses(input, TOKEN, (m, sink) => {
            if (m[1]) sink.emit(phonemizeWord(nat(m[1])));
            else if (m[2]) {
                // ≤9 digits fits a safe integer (<1e9) → compose; longer → read the raw string digit-by-digit so the
                // float conversion can't lose precision or go exponential (1e21 → "1e+21").
                const words = m[2].length <= 9 ? numberToWords(Number(m[2]), m[2]) : readDigits(m[2]);
                for (const wd of words.split(" ")) sink.emit(phonemizeWord(wd));
            }
            else if (m[3]) {
                const mk = CLAUSE_MARK[m[3]];
                if (mk) sink.pause(mk);
            }
        });
    }
}

/** Build the Finnish phonemizer (greedy g2p + gemination + velar-nasal rules + cardinal numbers). */
export function createFinnish(): Phonemizer {
    return new FinnishPhonemizer();
}

File: csharp/Vernacula.Phonemizer/Languages/Finnish/Finnish.cs
/**
 * Finnish (fi) phonemizer — Standard Finnish (yleiskieli), the Latin orthography, canonical IPA. One of the
 * most PHONEMICALLY TRANSPARENT orthographies there is, so this is a greedy longest-match scan over the
 * grapheme table with three code rules: CONSONANT GEMINATION (a doubled consonant → [Cː]) and the
 * velar-nasal pair ⟨ng⟩→ŋː (a LONG velar nasal) and ⟨nk⟩→ŋk. Consonant GRADATION is already spelled out in
 * the orthography, and fixed word-initial stress is predictable and unwritten → not emitted.
 * Ported from src/languages/finnish/finnish.ts — see that file for the tier's field-by-field argument.
 */
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Finnish;

public static class FinnishPhonemizer
{
    private static IReadOnlyDictionary<string, string> G => Manifest.MANIFEST.Graphemes;
    private static IReadOnlyDictionary<string, string> CLAUSE_MARK => Manifest.MANIFEST.ClausePunctuation;
    private static readonly HashSet<string> VOWEL_LETTERS = new(Manifest.MANIFEST.VowelLetters);

    /** Phonemize a single Finnish word to canonical IPA (segmental; gemination + the velar-nasal rules;
     *  length and diphthong offglides emitted). */
    public static string PhonemizeWord(string word)
    {
        var w = Js.ToLowerCase(word);
        var outp = new System.Text.StringBuilder();
        var i = 0;
        while (i < w.Length)
        {
            var c = w[i].ToString();
            var next = i + 1 < w.Length ? w[i + 1].ToString() : "";
            // ⟨ng⟩ → ŋː, consuming both letters — before gemination so the ⟨n⟩ is not mishandled.
            if (c == "n" && next == "g") { outp.Append("ŋː"); i += 2; continue; }
            // ⟨nk⟩ → ŋk: emit ŋ and let the k be scanned next, so kk-gemination still holds.
            if (c == "n" && next == "k") { outp.Append('ŋ'); i += 1; continue; }
            // consonant gemination: a doubled consonant letter → geminate [Cː].
            if (!VOWEL_LETTERS.Contains(c) && next == c && G.TryGetValue(c, out var gem) && gem.Length > 0)
            {
                outp.Append(gem).Append('ː');
                i += 2;
                continue;
            }
            var matched = false;
            foreach (var key in Manifest.GRAPHEME_KEYS)
            {
                if (i + key.Length <= w.Length && string.CompareOrdinal(w, i, key, 0, key.Length) == 0)
                {
                    outp.Append(G[key]);
                    i += key.Length;
                    matched = true;
                    break;
                }
            }
            // ⚠ NOT SILENTLY: a letter with no grapheme rule here still denotes a sound. Consulted only on
            // the MISS branch, after every grapheme has been tried.
            if (!matched)
            {
                outp.Append(LatinPhones.LatinPhone(w[i].ToString(), new PhoneOpts { Initial = i == 0, IncludeH = true }) ?? "");
                i += 1;
            }
        }
        return outp.ToString();
    }

    // A word (Finnish Latin letters incl. ä ö å + loan š ž) / number / punctuation token.
    private static readonly JsRe TOKEN = JsRegex.Compile($"({HostWord.LATIN_RUN})|(\\d+)|([.!?…,;:])", "giu");

    /** This language's OWN inventory — a different question from TOKEN's script boundary above. */
    private const string NATIVE_CLASS = "[a-zäöåšž]";
    private static readonly Func<string, string> Nat = HostWord.MakeNativiser(NATIVE_CLASS, "iu");

    /**
     * The shared symbol tier. Runs AFTER Normalize.cs, which is what lets that file keep its rewrites in
     * DIGITS — `13,6 cm` leaves it as `13 pilkku 6 cm`, so the number–unit adjacency is intact.
     *
     * ⚠ COUNT FORMS ARE `[singular, partitive]` AND THE DEFAULT SELECTOR IS EXACTLY RIGHT FOR FINNISH: a
     * counted noun is nominative after 1 and PARTITIVE after everything else.
     * ⚠ `km/h` AND `m/s` ARE COMPOUND KEYS, NOT `UnitPer` — the Finnish rate idiom takes the INESSIVE on the
     * denominator with no joining word at all, and `UnitPer` is one invariant string.
     * ⚠ BARE `m` IS DECLARED and the one-letter hazard was measured first; the residual exposure is a dotted
     * designation, which the tier's `NOT_VERSION` guard rejects by SEEING THE DOT — and this language's
     * normalizer never spends a decimal POINT (Finnish decimals use a comma), so the dot is still there.
     * ⚠ `°C` IS NOT ON THE TIER — Normalize.cs step 10 owns it, because it also reads the BARE `°`.
     */
    private static readonly Func<string, string> SYMBOLS = NormalizeSymbols.MakeSymbolNormalizer(new SymbolData
    {
        Percent = new[] { "prosentti", "prosenttia" },
        Currency = new Dictionary<string, IReadOnlyList<string>>
        {
            ["$"] = new[] { "dollari", "dollaria" }, ["€"] = new[] { "euro", "euroa" },
        },
        Units = new Dictionary<string, IReadOnlyList<string>>
        {
            // Longest-first inside the tier; both rate keys compose the inessive denominator themselves.
            ["km/h"] = new[] { "kilometri tunnissa", "kilometriä tunnissa" },
            ["m/s"] = new[] { "metri sekunnissa", "metriä sekunnissa" },
            ["km"] = new[] { "kilometri", "kilometriä" },
            ["cm"] = new[] { "senttimetri", "senttimetriä" },
            ["mm"] = new[] { "millimetri", "millimetriä" },
            ["kg"] = new[] { "kilogramma", "kilogrammaa" },
            ["m"] = new[] { "metri", "metriä" },
        },
        // Finnish welds the measure word onto the FRONT as one compound — *neliökilometriä* — which is
        // `compound`, never `before` (that would give two tokens).
        ExponentWords = new ExponentWordsDef
        {
            Squared = new[] { "neliö" }, Cubed = new[] { "kuutio" }, Position = ExponentPosition.Compound,
        },
        // The magnitude hop. Finnish takes NO connective, so `MagnitudeConnective` stays undeclared.
        Magnitudes = new[] { "miljoona", "miljoonaa", "miljardi", "miljardia", "biljoona", "biljoonaa" },
        Ampersand = "ja",
        // The corpus's `×` is a dimension cross and a multiplier; Finnish says *kertaa* for both, so `by`
        // defaults to `times`.
        Multiply = new MultiplyDef { Times = "kertaa" },
    });

    private sealed class Engine : ILanguage
    {
        public string Text(string rawInput)
        {
            var input = SYMBOLS(Normalize.NormalizeFinnishInitialisms(Normalize.NormalizeFinnish(rawInput)));
            return Clauses.AssembleClauses(input, TOKEN, (m, sink) =>
            {
                if (m.Groups[1].Success && m.Groups[1].Value.Length > 0)
                    sink.Emit(PhonemizeWord(Nat(m.Groups[1].Value)));
                else if (m.Groups[2].Success && m.Groups[2].Value.Length > 0)
                {
                    // ≤9 digits fits a safe integer → compose; longer → read the RAW STRING digit-by-digit so
                    // the float conversion cannot lose precision or go exponential.
                    var tok = m.Groups[2].Value;
                    var words = tok.Length <= 9 ? Numbers.NumberToWords(Js.Number(tok), tok) : Numbers.ReadDigits(tok);
                    foreach (var wd in words.Split(' ')) sink.Emit(PhonemizeWord(wd));
                }
                else if (m.Groups[3].Success && m.Groups[3].Value.Length > 0)
                {
                    if (CLAUSE_MARK.TryGetValue(m.Groups[3].Value, out var mk) && mk.Length > 0) sink.Pause(mk);
                }
            });
        }
    }

    /** Build the Finnish phonemizer (greedy g2p + gemination + the velar-nasal rules + cardinal numbers). */
    public static ILanguage CreateFinnish() => new Engine();

    internal static void RegisterSelf() => Registry.Register("finnish", () => CreateFinnish());
}
